File: May-LeetCoding-Challenge/24-Construct-Binary-Search-Tree-From-Preorder-Traversal/Construct-Binary-Search-Tree-From-Preorder-Traversal.cpp
class Solution {
public:
    TreeNode* construct(int i, int j, vector<int>& preorder){
        if (i > j) return NULL;
        TreeNode* root = new TreeNode(preorder[i]);
        int k = i;
        while (k <= j and preorder[k] <= preorder[i]) k++;
        root->left = construct(i+1, k-1, preorder);
        root->right = construct(k, j, preorder);
        return root;
    }
    
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        return construct(0, preorder.size() - 1, preorder);
    }
};

class Solution2 {
public:
    /* iterative, time : ~O(n log logn), space : O(n) */
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        TreeNode **v = new TreeNode*[preorder.size()];
        int curr_val, root_val = preorder[0];
        int pos = 0;
        TreeNode *root = new TreeNode(root_val);
        TreeNode *node = root;
        v[0] = root;
        
        for (int i = 1; i < preorder.size() ; ++i) {
            curr_val = preorder[i];
            if (preorder[i-1] > curr_val) {
                node->left = new TreeNode(curr_val);
                node = node->left;
                v[++pos] = node;
            } else { /* find the target father node */
                if (curr_val > root_val) {
                    pos = 0;
                    root_val = curr_val;
                } else {
                    /* interpolation search, do more time if data is large */
                    pos = ((curr_val - root_val) * (pos)) / (v[pos]->val - root_val);
                    pos = max(0, pos);
                    
                    /* target is near pos */
                    if (v[pos]->val > curr_val) {
                        while (v[++pos]->val > curr_val);
                    } else {
                        while (curr_val > v[--pos]->val);
                        ++pos;
                    }
                }
                node = new TreeNode(curr_val);
                v[pos]->right = node;
                v[pos] = node;
            }
        }
        return root;
    }
};

File: May-LeetCoding-Challenge/24-Construct-Binary-Search-Tree-From-Preorder-Traversal/Construct-Binary-Search-Tree-From-Preorder-Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * Recursive solution, construct the left tree and right tree recursively, and the combine to the root.
     * @param preorder
     * @return
     */
    public TreeNode bstFromPreorder(int[] preorder) {
        if(preorder.length == 0) return null;
        
        int rootVal = preorder[0];
        if(preorder.length == 1) return new TreeNode(rootVal);
        
        TreeNode rightTree = null;
        TreeNode leftTree = null;
        TreeNode root = new TreeNode(rootVal, null, null);
        
        if(preorder[1] > rootVal) {
            // if there is no left tree
            rightTree = bstFromPreorder(Arrays.copyOfRange(preorder, 1, preorder.length));
        } else {
            int i = 1;
            while (i < preorder.length) {
                // find separate index of left tree and right tree
                if (preorder[i] > rootVal) {
                    break;
                }
                i++;
            }
            
            // construct left tree
            leftTree = bstFromPreorder(Arrays.copyOfRange(preorder, 1, i));
            if(i != preorder.length) {
                //construct right tree
                rightTree = bstFromPreorder(Arrays.copyOfRange(preorder, i, preorder.length));
            }
        }        
        
        root.left = leftTree;
        root.right = rightTree;
        
        return root;
    }
}